File: Lista Doble Circular/DobleCircular.cs
using System;

public class Node
{
    public int Data;
    public Node Next;
    public Node Prev;
}

public class DoublyCircularLinkedList
{
    private Node head;
    private Node tail;

    public DoublyCircularLinkedList()
    {
        head = null;
        tail = null;
    }

    public void InsertBegin()
    {
        Console.Write("Ingrese un valor: ");
        int value = int.Parse(Console.ReadLine());
        Node newNode = new Node();
        newNode.Data = value;

        if (head == null)
        {
            head = newNode;
            tail = newNode;
            newNode.Next = newNode;
            newNode.Prev = newNode;
        }
        else
        {
            newNode.Next = head;
            newNode.Prev = tail;
            head.Prev = newNode;
            tail.Next = newNode;
            head = newNode;
        }
    }

    public void InsertEnd()
    {
        Console.Write("Ingrese un valor: ");
        int value = int.Parse(Console.ReadLine());
        Node newNode = new Node();
        newNode.Data = value;

        if (head == null)
        {
            head = newNode;
            tail = newNode;
            newNode.Next = newNode;
            newNode.Prev = newNode;
        }
        else
        {
            newNode.Next = head;
            newNode.Prev = tail;
            tail.Next = newNode;
            head.Prev = newNode;
            tail = newNode;
        }
    }

    public void InsertRandom()
    {
        Console.Write("Ingrese un valor: ");
        int value = int.Parse(Console.ReadLine());
        Console.Write("Ingrese la posicion donde desea insertar: ");
        int pos = int.Parse(Console.ReadLine());

        Node newNode = new Node();
        newNode.Data = value;

        if (head == null)
        {
            head = newNode;
            tail = newNode;
            newNode.Next = newNode;
            newNode.Prev = newNode;
        }
        else
        {
            Node current = head;
            for (int i = 0; i < pos; i++)
            {
                current = current.Next;
                if (current == head) break;
            }
            newNode.Next = current;
            newNode.Prev = current.Prev;
            current.Prev.Next = newNode;
            current.Prev = newNode;

            if (current == head && pos == 0) head = newNode;
            if (current == head) tail = newNode;
        }
    }

    public void DeleteBegin()
    {
        if (head == null)
        {
            Console.WriteLine("La lista esta vacia");
            return;
        }
        if (head == tail)
        {
            head = null;
            tail = null;
        }
        else
        {
            Node temp = head;
            head = head.Next;
            head.Prev = tail;
            tail.Next = head;
        }
        Console.WriteLine("Elemento eliminado desde el inicio");
    }

    public void DeleteEnd()
    {
        if (head == null)
        {
            Console.WriteLine("La lista esta vacia");
            return;
        }
        if (head == tail)
        {
            head = null;
            tail = null;
        }
        else
        {
            Node temp = tail;
            tail = tail.Prev;
            tail.Next = head;
            head.Prev = tail;
        }
        Console.WriteLine("Elemento eliminado desde el final");
    }

    public void DeleteRandom()
    {
        if (head == null)
        {
            Console.WriteLine("La lista esta vacia");
            return;
        }
        Node current = head;
        Console.Write("Ingresa la posicion: ");
        int pos = int.Parse(Console.ReadLine());
        for (int i = 0; i < pos; i++)
        {
            current = current.Next;
            if (current == head)
            {
                Console.WriteLine("La posicion no existe");
                return;
            }
        }
        current.Prev.Next = current.Next;
        current.Next.Prev = current.Prev;

        if (current == head) head = current.Next;
        if (current == tail) tail = current.Prev;
        if (head == null) tail = null;
        Console.WriteLine($"Elemento eliminado desde la posicion {pos}");
    }

    public void Search()
    {
        if (head == null)
        {
            Console.WriteLine("La lista esta vacia");
            return;
        }
        Console.Write("Ingrese el valor a buscar: ");
        int value = int.Parse(Console.ReadLine());
        Node temp = head;
        int pos = 0;
        do
        {
            if (temp.Data == value)
            {
                Console.WriteLine($"Valor {value} encontrado en la posicion {pos}");
                return;
            }
            temp = temp.Next;
            pos++;
        } while (temp != head);
        Console.WriteLine($"Valor {value} no encontrado en la lista");
    }

    public void Display()
    {
        if (head == null)
        {
            Console.WriteLine("La lista esta vacia");
            return;
        }
        Node temp = head;
        Console.Write("Lista Doble Circular: ");
        do
        {
            Console.Write($"{temp.Data} <-> ");
            temp = temp.Next;
        } while (temp != head);
        Console.WriteLine("inicio");
    }
}

public class Program
{
    public static void Main()
    {
        DoublyCircularLinkedList dcll = new DoublyCircularLinkedList();
        int choice;
        do
        {
            Console.WriteLine("1. Insertar al inicio");
            Console.WriteLine("2. Insertar al final");
            Console.WriteLine("3. insertar en posicion aleatoria");
            Console.WriteLine("4. Eliminar al inicio");
            Console.WriteLine("5. Eliminar al final");
            Console.WriteLine("6. Eliminar desde posicion especifica");
            Console.WriteLine("7. Buscar");
            Console.WriteLine("8. Mostrar lista");
            Console.WriteLine("9. Salir");
            Console.Write("Ingrese su opcion: ");
            choice = int.Parse(Console.ReadLine());

            switch (choice)
            {
                case 1: dcll.InsertBegin(); break;
                case 2: dcll.InsertEnd(); break;
                case 3: dcll.InsertRandom(); break;
                case 4: dcll.DeleteBegin(); break;
                case 5: dcll.DeleteEnd(); break;
                case 6: dcll.DeleteRandom(); break;
                case 7: dcll.Search(); break;
                case 8: dcll.Display(); break;
                case 9: Console.WriteLine("Saliendo..."); break;
                default: Console.WriteLine("Opcion invalida. Intente de nuevo.");
            }
        } while (choice != 9);
    }
}

File: Lista Doble Circular/DobleCircular.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
    Node* prev;
};

class DoublyCircularLinkedList {
private:
    Node* head;
    Node* tail; 
public:
    DoublyCircularLinkedList() : head(nullptr), tail(nullptr) {}
    void insertbegin() {
        int value;
        cout << "Ingrese un valor: ";
        cin >> value;

        Node* newNode = new Node();
        newNode->data = value;

        if (head == nullptr) {
            head = newNode;
            tail = newNode;
            newNode->next = newNode;
            newNode->prev = newNode;
        } else {
            newNode->next = head;
            newNode->prev = tail;
            head->prev = newNode;
            tail->next = newNode;
            head = newNode;
        }
    }
    void insertend(){
        int value;
        cout << "Ingrese un valor: ";
        cin >> value;

        Node* newNode = new Node();
        newNode->data = value;

        if (head == nullptr) {
            head = newNode;
            tail = newNode;
            newNode->next = newNode;
            newNode->prev = newNode;
        } else {
            newNode->next = head;
            newNode->prev = tail;
            tail->next = newNode;
            head->prev = newNode;
            tail = newNode;
        }
    }
    void insertrandom(){
        int value;
        int pos;
        cout << "Ingrese un valor: ";
        cin >> value;
        cout << "Ingrese la posicion donde desea insertar: ";
        cin >> pos;

        Node* newNode = new Node();
        newNode->data = value;

        if (head == nullptr) {
            head = newNode;
            tail = newNode;
            newNode->next = newNode;
            newNode->prev = newNode;
        } else {
            Node* current = head;
            for (int i = 0; i < pos; i++) {
                current = current->next;
                if (current == head) {
                    break; 
                }
            }
            newNode->next = current;
            newNode->prev = current->prev;
            current->prev->next = newNode;
            current->prev = newNode;

            if (current == head && pos == 0) {
                head = newNode; 
            }
            if (current == head) {
                tail = newNode; 
            }
        }
    }
    void deletebegin(){
        if (head == nullptr) {
            cout << "La lista esta vacia" << endl;
            return;
        }

        if (head == tail) {
            delete head;
            head = nullptr;
            tail = nullptr;
        } else {
            Node* temp = head;
            head = head->next;
            head->prev = tail;
            tail->next = head;
            delete temp;
        }
        cout<< "Elemento eliminado desde el inicio" << endl;
    }
    void deleteend(){
        if (head == nullptr) {
            cout << "La lista esta vacia" << endl;
            return;
        }

        if (head == tail) {
            delete head;
            head = nullptr;
            tail = nullptr;
        } else {
            Node* temp = tail;
            tail = tail->prev;
            tail->next = head;
            head->prev = tail;
            delete temp;
        }
        cout<< "Elemento eliminado desde el final" << endl;
    }
    void deleteRandom(){
        if (head == nullptr) {
            cout << "La lista esta vacia" << endl;
            return;
        }
        Node* current = head;
        int pos;
        cout << "Ingresa la posicion: ";
        cin >> pos;
        for (int i = 0; i < pos; i++) {
            current = current->next;
            if (current == head) {
                cout << "La posicion no existe" << endl;
                return;
            }
        }
        current->prev->next = current->next;
        current->next->prev = current->prev;

        if (current == head) {
            head = current->next; 
        }
        if (current == tail) {
            tail = current->prev; 
        }
        if (head == nullptr) {
            tail = nullptr; 
        }
        delete current;
        cout << "Elemento eliminado desde la posicion " << pos << endl;
    }
    void search(){
        if (head == nullptr) {
            cout << "La lista esta vacia" << endl;
            return;
        }
        int value;
        cout << "Ingrese el valor a buscar: ";
        cin >> value;

        Node* temp = head;
        int pos = 0;
        do {
            if (temp->data == value) {
                cout << "Valor " << value << " encontrado en la posicion " << pos << endl;
                return;
            }
            temp = temp->next;
            pos++;
        } while (temp != head);
        cout << "Valor " << value << " no encontrado en la lista" << endl;
    }
    void display() {
        if (head == nullptr) {
            cout << "La lista esta vacia" << endl;
            return;
        }
        Node* temp = head;
        cout << "Lista Doble Circular: ";
        do {
            cout << temp->data << " <-> ";
            temp = temp->next;
        } while (temp != head);
        cout <<"inicio"<< endl;
    }
};

int main() {
    DoublyCircularLinkedList dcll;
    int choice;

    do {
        cout << "1. Insertar al inicio" << endl;
        cout << "2. Insertar al final" << endl;
        cout << "3. insertar en posicion aleatoria" << endl;
        cout << "4. Eliminar al inicio" << endl;
        cout << "5. Eliminar al final" << endl;
        cout << "6. Eliminar desde posicion especifica" << endl;
        cout << "7. Buscar" << endl;
        cout << "8. Mostrar lista" << endl;
        cout << "9. Salir" << endl;
        cout << "Ingrese su opcion: ";
        cin >> choice;

        switch (choice) {
            case 1:
                dcll.insertbegin();
                break;
            case 2:
                dcll.insertend();
                break;
            case 3:
                dcll.insertrandom();
                break;
            case 4:
                dcll.deletebegin();
                break;
            case 5:
                dcll.deleteend();
                break;
            case 6:
                dcll.deleteRandom();
                break;
            case 7:
                dcll.search();
                break;
            case 8:
                dcll.display();
                break;
            case 9:
                cout << "Saliendo..." << endl;
                break;
            default:
                cout << "Opcion invalida. Intente de nuevo." << endl;
        }
    } while (choice != 9);

    return 0;
}